src/libsyntax/parse/lexer/mod.rs

```diff
@@ -292,28 +243,369 @@ impl<'a> StringReader<'a> {
         self.sess.span_diagnostic.struct_span_fatal(self.mk_sp(from_pos, to_pos), &m[..])
     }
 
-    /// Advance peek_token to refer to the next token, and
+    /// Advance peek_tok and peek_span to refer to the next token, and
     /// possibly update the interner.
     fn advance_token(&mut self) -> Result<(), ()> {
-        match self.scan_whitespace_or_comment() {
-            Some(comment) => {
-                self.peek_span_src_raw = comment.span;
-                self.peek_token = comment;
+        if self.is_eof() {
+            let (real, raw) = self.mk_sp_and_raw(
+                self.source_file.end_pos,
+                self.source_file.end_pos
+            );
+            self.peek_token = Token::new(token::Eof, real);
+            self.peek_span_src_raw = raw;
+            return Ok(());
+        }
+
+        let text: &str = &self.src[(self.pos - self.source_file.start_pos).to_usize()..];
+        debug_assert!(!text.is_empty());
+
+        {
+            //FIXME: move shebang parsing out of the inner loop
+            let is_beginning_of_file = self.pos == self.source_file.start_pos;
+            if is_beginning_of_file {
+                if let Some(shebang_len) = rustc_lexer::strip_shebang(text) {
+                    self.bump_bytes("#!".len());
+                    let name_start = self.pos;
+                    self.bump_bytes(shebang_len - "#!".len());
+                    let sp = self.mk_sp(name_start, self.pos);
+                    self.peek_token = Token::new(token::Shebang(self.name_from(name_start)), sp);
+                    self.peek_span_src_raw = sp;
+                    return Ok(());
+                }
+            }
+        }
+
+        let token = rustc_lexer::first_token(text);
+
+        let start = self.pos;
+        self.bump_bytes(token.len);
+        let kind = self.lexer_token_to_ast_token(token.kind, start)
+            .map_err(|err| self.fatal_errs.push(err))?;
+
+        let (real, raw) = self.mk_sp_and_raw(start, self.pos);
+        self.peek_span_src_raw = raw;
+        self.peek_token = Token::new(kind, real);
+        Ok(())
+    }
+
+    fn lexer_token_to_ast_token(
```

Comment:
This is rather a low-level lexer token into language-level lexer token, the AST is still far away.

Language lexer's behavior is a part of the language since it determines what can be passed as tokens to proc macros and what not.
It's kind of unfortunate to introduce these two layers - you have to have some really good understanding of what the low-level lexer accepts and what recovery it performs silently to correctly report the subset that is not supposed to be accepted by the language lexer.

Reply:
For example, lifetime names starting with a number.
It's a pretty arbitrary recovery technique that makes sense if you check the produced lifetime name, but becomes risky when you have the low-level lexer and error detection recoverable.
For all other tools using results produced by the low-level lexer `'0abc` is going to look like a normal lifetime token.

Reply:
This can probably be improved by giving error condition flags (like `terminated` for some literals) to numeric lifetimes and other tokens that perform silently perform recovery from the language lexer's point of view.

Reply:
On the other hand the low-level lexer probably shouldn't leave flags for things like incorrect escapes in string literals, that... well, don't mess with the lexical structure of the token like numeric lifetimes do.
I'm not sure where the exact border lies.

Reply:
Yeah, this is a valid concern. I guess, the rule of thumb should be that the code that does recovery also should be responsible for setting an error flag, but doesn't have to be concerned with good error reporting. Specifically, lifetime starting with number should have error flag set.

I also think that additional validation, like the `unescape` module, also belongs to the rustc_lexer crate, it just shouldn't be intertwined with the lexing process itself.